api/services/S3Helper.js

```diff
@@ -0,0 +1,128 @@
+const AWS = require('aws-sdk');
+const config = require('../../config');
+
+const s3Config = config.s3;
+
+const S3_DEFAULT_MAX_KEYS = 1000;
+
+const s3Client = new AWS.S3({
+  accessKeyId: s3Config.accessKeyId,
+  secretAccessKey: s3Config.secretAccessKey,
+  region: s3Config.region,
+});
+
+function listObjectsHelper(currObjects, extraS3Params = {}, opts = {}, callback) {
+  const listObjectArgs = Object.assign({}, {
+    Bucket: s3Config.bucket,
+  }, extraS3Params);
+
+  s3Client.listObjectsV2(listObjectArgs, (err, data) => {
+    if (err) {
+      callback(err);
+      return;
+    }
+
+    const aggregationKey = opts.aggregationKey || 'Contents';
+    const totalMaxObjects = opts.totalMaxObjects;
+
+    const objects = currObjects ?
+      currObjects.concat(data[aggregationKey])
+      : data[aggregationKey];
+
+    // if results should be limited
+    if (totalMaxObjects) {
+      // We're ready to exit if either the response data are not truncated
+      // or the current objects length is >= to the desired totalMaxObjects
+      if (!data.IsTruncated || objects.length >= totalMaxObjects) {
+        // then we should return early
+
+        // first, truncate the objects to the maximum total objects
+        // or its length (slice will only go as far as the array's length)
+        const truncatedObjects = objects.slice(0, totalMaxObjects);
+
+        const pagedResults = {
+          isTruncated: data.IsTruncated,
+          objects: truncatedObjects,
+        };
+
+        callback(null, pagedResults);
+        return;
+      }
+      // otherwise continue as normal (ie, not paged)
+    }
+
+    if (data.IsTruncated) {
+      const newExtraParams = Object.assign({},
+        extraS3Params,
+        { ContinuationToken: data.NextContinuationToken }
+      );
+      // call recursively
+      listObjectsHelper(objects, newExtraParams, opts, callback);
+    } else {
+      // done !
+      callback(null, objects);
+    }
+  });
+}
+
+function resolveCallback(resolve, reject) {
```

Comment:
Why this construct vs wrapping the listObjectHelper  function in a promise? I suppose this is easier to think about than a recursive promise.

Reply:
yeah, that's the reason. I found the callback easier to think about with the recursion.